pkg/eventmanager/eventmanager.go

```diff
@@ -0,0 +1,224 @@
+	if err != nil {
+		return fmt.Errorf("can't check cluster metric ClusterVersion: %v", err)
+	} else {
+		upgradeStarted = isSet
+	}
+
+	// As a backup if metrics is unavailable, check if upgradeConfig indicates the upgrade has completed
+	if !upgradeStarted {
+		upgradePhase, err := getUpgradePhase(uc)
+		if err == nil && *upgradePhase == upgradev1alpha1.UpgradePhaseUpgraded {
+			upgradeStarted = true
+		}
+	}
+
+	if !upgradeStarted {
+		return nil
+	}
+
+	// Check if a notification for it has already been sent successfully
+	isNotified, err := mc.IsMetricNotificationEventSentSet(uc.Name, string(notifier.StateStarted), uc.Spec.Desired.Version)
+	if err != nil {
+		return fmt.Errorf("can't check cluster metric NotificationSent: %v", err)
+	}
+	if isNotified {
+		return nil
+	}
+
+	// Send the notification and indicate it has been sent
+	description := fmt.Sprintf("Cluster is currently being upgraded to version %s", uc.Spec.Desired.Version)
+	err = nc.NotifyState(notifier.StateStarted, description)
+	if err != nil {
+		return fmt.Errorf("can't send notification '%s': %v", notifier.StateStarted, err)
+	}
+	mc.UpdateMetricNotificationEventSent(uc.Name, string(notifier.StateStarted), uc.Spec.Desired.Version)
+
+	return nil
+}
+
+func checkUpgradeEnd(mc metrics.Metrics, nc notifier.Notifier, uc *upgradev1alpha1.UpgradeConfig) error {
+
+	upgradeEnded := false
+
+	// Check metrics if the node upgrade end time metric has been set
+	isSet, err := mc.IsMetricNodeUpgradeEndTimeSet(uc.Name, uc.Spec.Desired.Version)
+	if err != nil {
+		log.Error(err, "could not check metrics for upgrade end time")
+	} else {
+	   upgradeEnded = isSet
+	}
+
+	// As a backup, check if upgradeConfig indicates the upgrade has completed
+	if !upgradeEnded {
+		upgradePhase, err := getUpgradePhase(uc)
+		if err == nil && *upgradePhase == upgradev1alpha1.UpgradePhaseUpgraded {
+			upgradeEnded = true
+		}
+	}
+
+	// If the upgrade hasn't ended, do nothing
+	if !upgradeEnded {
+		return nil
+	}
+
+	// Check if a notification for it has been sent successfully
+	isNotified, err := mc.IsMetricNotificationEventSentSet(uc.Name, string(notifier.StateCompleted), uc.Spec.Desired.Version)
+	if err != nil {
+		return fmt.Errorf("can't check cluster metric NotificationSent: %v", err)
+	}
+	if isNotified {
+		return nil
+	}
+
+	// Send the notification and indicate it has been sent
+	description := fmt.Sprintf("Cluster has been successfully upgraded to version %s", uc.Spec.Desired.Version)
+	err = nc.NotifyState(notifier.StateCompleted, description)
+	if err != nil {
+		return fmt.Errorf("can't send notification '%s': %v", notifier.StateCompleted, err)
+	}
+	mc.UpdateMetricNotificationEventSent(uc.Name, string(notifier.StateCompleted), uc.Spec.Desired.Version)
+
+	return nil
+}
+
+// Returns the phase of the currently executing upgrade or error if no phase can be found
+func getUpgradePhase(uc *upgradev1alpha1.UpgradeConfig) (*upgradev1alpha1.UpgradePhase, error) {
```

Comment:
Just wondering should this be in the upgradeconfigmanager?

Reply:
I think it definitely deserves being elsewhere - my thoughts were maybe as part of one of the UpgradeConfig types, rather than in the UC Manager. It could sit alongside GetHistory() and functions like that.